app/packages/core/src/components/Grid/useExpandSample.ts

```diff
@@ -1,47 +1,112 @@
-import { useCallback } from "react";
+import { Snapshot, useRecoilCallback } from "recoil";
 
-import { FlashlightConfig } from "@fiftyone/flashlight";
+import { ItemIndexMap } from "@fiftyone/flashlight/src/state";
 import * as fos from "@fiftyone/state";
+import { getFetchFunction } from "@fiftyone/utilities";
+
+import { pageParameters } from "./recoil";
+
+interface SamplesResponse {
+  results: Array<{
+    sample: fos.AppSample & { frame_number?: number };
+    aspect_ratio: number;
+    frame_rate?: number;
+    urls: Array<{ field: string; url: string }>;
+    thumbnails_only: boolean;
+  }>;
+}
+
+/**
+ * Returns a full sample by id, querying from the backend if necessary.
+ *
+ * Throws if the provided sampleId does not exist in the store.
+ */
+async function getFullSample<T extends fos.Lookers>(
+  store: fos.LookerStore<T>,
+  snapshot: Snapshot,
+  sampleId: string
+): Promise<fos.SampleData | null> {
+  const existingSample = store.samples.get(sampleId);
+
+  if (!existingSample) {
+    return null;
+  }
+
+  if (!existingSample.thumbnailsOnly) {
+    return existingSample;
+  } else {
+    // Query for the full sample from the backend
+    const { zoom, ...params } = await snapshot.getPromise(
+      pageParameters(/* modal= */ true)
+    );
+
+    const { results } = (await getFetchFunction()("POST", "/samples", {
```

Comment:
This bit of code is similar to the block here: https://github.com/compound-eye/fiftyone/blob/0b542e5bda3308a46b157367dfa3b1ebff7824b2/app/packages/core/src/components/Grid/usePage.ts#L26